File: c++/baekjoon/1987.cpp
#include <iostream>
#include <set>

using namespace std;
char board[20][20];
int cnt, max_cnt;
set<char> alphabet;
int R, C;

void dfs(int i, int j);

int main(){
    cnt = max_cnt = 0;
    cin >> R >> C;

    for(int i=0;i<R;i++){
        cin >> board[i];
    }

    dfs(0, 0);

    cout << max_cnt;
}

void dfs(int i, int j){
    alphabet.insert(board[i][j]);
    cnt++;
    if(max_cnt < cnt){
        max_cnt = cnt;
    }
    if(i > 0 && alphabet.find(board[i-1][j]) == alphabet.end()){
        dfs(i-1, j);
    }
    if(i < R-1 && alphabet.find(board[i+1][j]) == alphabet.end()){
        dfs(i+1, j);
    }
    if(j > 0 && alphabet.find(board[i][j-1]) == alphabet.end()){
        dfs(i, j-1);
    }
    if(j < C-1 && alphabet.find(board[i][j+1]) == alphabet.end()){
        dfs(i, j+1);
    }

    cnt--;
    alphabet.erase(alphabet.find(board[i][j]));
}

File: python/baekjoon/1987.py
# Created by Jello on 2017. 7. 5.
def dfs(i, j):
    global count, max_count
    alphabet.add(board[i][j])
    count += 1
    max_count = max(max_count, count)
    print('Add {}, count={}, max_count={}'.format(board[i][j], count, max_count))

    if i > 0 and board[i-1][j] not in alphabet:
        dfs(i-1, j)
    if i < R-1 and board[i+1][j] not in alphabet:
        dfs(i+1, j)
    if j > 0 and board[i][j-1] not in alphabet:
        dfs(i, j-1)
    if j < C-1 and board[i][j+1] not in alphabet:
        dfs(i, j+1)

    alphabet.remove(board[i][j])
    count -= 1
    print('{} removed'.format(board[i][j]))


R, C = map(int, str(input()).split(' '))
board = [[] for x in range(R)]
alphabet = set()
count = max_count = 0

for i in range(R):
    board[i] = str(input())

# DFS
dfs(0, 0)

print(max_count)
